src/main/java/io/jenkins/plugins/folderauth/FolderBasedAuthorizationStrategy.java

```diff
@@ -504,6 +506,50 @@ private void initCache() {
                 .build();
     }
 
+    /**
+     * Call this function when a folder gets deleted from Jenkins.
+     *
+     * @param fullName the full name of the {@link com.cloudbees.hudson.plugins.folder.AbstractFolder}.
+     */
+    @Restricted(NoExternalUse.class)
+    public void onFolderDeleted(String fullName) {
+        folderRoles.parallelStream().forEach(role -> role.removeFolder(fullName));
+        updateJobAcls(true);
+        jobAclCache.invalidateAll();
+        try {
+            Jenkins.get().save();
+        } catch (IOException e) {
+            LOGGER.log(Level.SEVERE, "Unable to update folder roles and save configuration when folder "
+                                         + fullName + " was deleted.", e);
+        }
+    }
+
+    /**
+     * Call this function when a folder gets renamed or relocated in Jenkins.
+     *
+     * @param oldFullName the old full name of the {@link com.cloudbees.hudson.plugins.folder.AbstractFolder}
+     * @param newFullName the new full name of the {@link com.cloudbees.hudson.plugins.folder.AbstractFolder}
+     */
+    @Restricted(NoExternalUse.class)
+    public void onFolderRenamed(String oldFullName, String newFullName) {
+        folderRoles.parallelStream().forEach(role -> {
+            if (role.getFolderNames().contains(newFullName)) {
+                role.removeFolder(oldFullName);
```

Comment:
This operation does not look to be atomic. Any chance of race conditions?

Reply:
Race conditions *should* not take place here, the set of folder names used by the `FolderRole` is not modified apart from the added `addFolder()` and `removeFolder()` methods. They are invoked only when a folder is either deleted or renamed and are `@Restricted(NoExternalUse.class)`.